photon-ml/src/main/scala/com/linkedin/photon/ml/ModelTraining.scala

```diff
@@ -155,42 +160,34 @@ object ModelTraining extends Logging {
       s"${warmStartModels.keys.mkString(", ")}")
 
     val initWeightsAndModels = List[(Double, GeneralizedLinearModel)]()
-    val (finalOptimizationProblem, finalWeightsAndModels) = sortedRegularizationWeights
-      .foldLeft((initOptimizationProblem, initWeightsAndModels)) {
-        case ((optimizationProblem, List()), currentWeight) =>
+    val finalWeightsAndModels = sortedRegularizationWeights
+      .foldLeft(initWeightsAndModels) {
+        case (List(), currentWeight) =>
           // Initialize the list with the result from the first regularization weight
-          val updatedOptimizationProblem = optimizationProblem.updateObjective(
-            normalizationContextProvider, currentWeight)
+          optimizationProblem.updateRegularizationWeight(currentWeight)
```

Comment:
Yes, the update the regularization weight for the problem and re-run. It was easier to add in regularization weight updates as a change to state, and in line with optimization state tracking which is also stateful. However, as mentioned above, I'm not opposed to going through and making the design functional.